### PR TITLE
Add retry button in Pipeline UI

### PR DESCRIPTION
**Pull request by IronPan**
/assign @hongye-sun @rileyjbauer

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1782)

### CODE REVIEW

File: frontend/src/lib/Buttons.ts [hunk 1]
```diff
@@ -249,6 +261,19 @@ export default class Buttons {
     }
   }
 
+  private _retryRun(selectedIds: string[], useCurrent: boolean,
+                    callback: (selectedIds: string[], success: boolean) => void): void {
+    this._dialogActionHandler(
+        selectedIds,
+        `Retry this run?`,
+        useCurrent,
+        id => Apis.runServiceApi.retryRun(id),
```
Comment (ajchili) [inline frontend/src/lib/Buttons.ts]: I believe this can be changed to `Apis.runServiceApi.retryRun`

File: frontend/src/lib/Buttons.ts [hunk 2]
```diff
@@ -249,6 +261,19 @@ export default class Buttons {
     }
   }
 
+  private _retryRun(selectedIds: string[], useCurrent: boolean,
+                    callback: (selectedIds: string[], success: boolean) => void): void {
+    this._dialogActionHandler(
+        selectedIds,
+        `Retry this run?`,
```
Comment (ajchili) [inline frontend/src/lib/Buttons.ts]: Should ``` ` ``` be used rather than `'`? Will template strings be used?

**Comment by ajchili** [comment]
You also have to update snapshots for the UI unit tests. This can be done by running `npm test -- --watch` then pressing `u` after the tests fail. But this will also require additional updates to other existing tests that do not rely on snapshots.

File: frontend/src/pages/RunDetails.test.tsx [hunk 3]
```diff
@@ -849,7 +849,7 @@ describe('RunDetails', () => {
       });
     });
 
-    [NodePhase.PENDING, NodePhase.RUNNING, NodePhase.UNKNOWN].forEach(status => {
+    [NodePhase.ERROR, NodePhase.FAILED, NodePhase.PENDING, NodePhase.RUNNING, NodePhase.UNKNOWN, NodePhase.SKIPPED].forEach(status => {
```
[the remark this answers is not in the document]
Comment (IronPan) [reply frontend/src/pages/RunDetails.test.tsx]: done

File: frontend/src/pages/RunDetails.test.tsx [hunk 4]
```diff
@@ -599,7 +599,7 @@ describe('RunDetails', () => {
     expect(tree.state('selectedNodeDetails')).toHaveProperty('phaseMessage', undefined);
   });
 
-  [NodePhase.RUNNING, NodePhase.PENDING, NodePhase.UNKNOWN].forEach(unfinishedStatus => {
+  [NodePhase.ERROR, NodePhase.FAILED, NodePhase.RUNNING, NodePhase.PENDING, NodePhase.UNKNOWN, NodePhase.SKIPPED].forEach(unfinishedStatus => {
```
[the remark this answers is not in the document]
Comment (IronPan) [reply frontend/src/pages/RunDetails.test.tsx]: done

File: frontend/src/pages/RunDetails.test.tsx [hunk 5]
```diff
@@ -839,7 +839,7 @@ describe('RunDetails', () => {
     }, 10000);
 
 
-    [NodePhase.ERROR, NodePhase.FAILED, NodePhase.SUCCEEDED, NodePhase.SKIPPED].forEach(status => {
+    [ NodePhase.SUCCEEDED].forEach(status => {
```
[the remark this answers is not in the document]
Comment (IronPan) [reply frontend/src/pages/RunDetails.test.tsx]: done

File: frontend/src/lib/Buttons.ts [hunk 6]
```diff
@@ -258,7 +283,7 @@ export default class Buttons {
       `be stopped if it's running when it's archived. Use the Restore action to restore the ` +
       `run${s(selectedIds)} to ${selectedIds.length === 1 ? 'its' : 'their'} original location.`,
       useCurrent,
-      id => Apis.runServiceApi.archiveRun(id),
+      Apis.runServiceApi.archiveRun,
```
Comment (rileyjbauer) [inline frontend/src/lib/Buttons.ts]: why this change?

File: frontend/src/lib/StatusUtils.ts [hunk 7]
```diff
@@ -40,11 +40,11 @@ export enum NodePhase {
...
+      return true;
     case NodePhase.FAILED: // Fall through
     case NodePhase.ERROR: // Fall through
     case NodePhase.SKIPPED: // Fall through
-    case NodePhase.TERMINATED:
-      return true;
+    case NodePhase.TERMINATED: // Fall through
     case NodePhase.PENDING: // Fall through
     case NodePhase.RUNNING: // Fall through
     case NodePhase.TERMINATING: // Fall through
```
[the remark this answers is not in the document]
Comment (rileyjbauer) [reply frontend/src/lib/StatusUtils.ts]: No, @IronPan this is problematic with the existing use of `hasFinished`. This function is supposed to indicate that the the run had stopped running for any reason. For example, it is used in the `RunDetails` page to determine whether to continue autorefreshing because a `pending` run might update over time, but a `failed` run never will.

Is there any reason you can't create a separate function called `hasSucceeded` or something like that?

File: frontend/src/lib/Buttons.ts (hunk 6 above)
Comment (ajchili) [reply frontend/src/lib/Buttons.ts]: It was a style changed proposed by me, it can be reverted for readability.

Comment (rileyjbauer) [reply frontend/src/lib/Buttons.ts]: I believe there is actually a difference between using the arrow notation as opposed to just passing the function itself. The arrow notation captures the `this` of the surrounding scope whereas if just the function is passed, `this` is limited to the internal scope of the function

Comment (ajchili) [reply frontend/src/lib/Buttons.ts]: I hadn't considered that, I believe you are correct. This should be reverted as @rileyjbauer suggested.

File: frontend/src/lib/Buttons.ts (hunk 1 above)
Comment (IronPan) [reply frontend/src/lib/Buttons.ts]: done

File: frontend/src/lib/StatusUtils.test.tsx [hunk 8]
```diff
@@ -19,13 +19,13 @@ import { NodePhase, hasFinished, statusBgColors, statusToBgColor, checkIfTermina
 describe('StatusUtils', () => {
 
   describe('hasFinished', () => {
-    [NodePhase.ERROR, NodePhase.FAILED, NodePhase.SUCCEEDED, NodePhase.SKIPPED, NodePhase.TERMINATED].forEach(status => {
+    [ NodePhase.SUCCEEDED].forEach(status => {
```
[the remark this answers is not in the document]
Comment (IronPan) [reply frontend/src/lib/StatusUtils.test.tsx]: done

File: frontend/src/lib/StatusUtils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (IronPan) [reply frontend/src/lib/StatusUtils.ts]: The goal for this change is to continue to refresh on non succeeded runs because those might still change the status once user click retry.

File: frontend/src/components/Graph.tsx [hunk 9]
```diff
@@ -154,6 +154,9 @@ export default class Graph extends React.Component<GraphProps, GraphState> {
           // deviating from the explicit layout provided by dagre.
           if (i === 1) {
             const sourceNode = graph.node(edgeInfo.v);
+            if (sourceNode === undefined) {
```
Comment (rileyjbauer) [inline frontend/src/components/Graph.tsx]: I believe this (and below) can be removed with the fix to `WorkflowParser`

File: frontend/src/components/Graph.tsx [hunk 10]
```diff
@@ -176,10 +179,13 @@ export default class Graph extends React.Component<GraphProps, GraphState> {
...
           if (finalSegment) {
             const destinationNode = graph.node(edgeInfo.w);
+            if (destinationNode === undefined) {
+              continue;
+            }
 
             // Placeholder nodes never need adjustment because they always have only a single
             // incoming edge.
-            if (!destinationNode.isPlaceholder) {
+            if (destinationNode && !destinationNode.isPlaceholder) {
```
Comment (rileyjbauer) [inline frontend/src/components/Graph.tsx]: This can be removed, especially if the `destinationNode === undefined` is kept

File: frontend/src/components/Graph.tsx (hunk 9 above)
Comment (IronPan) [reply frontend/src/components/Graph.tsx]: done

File: frontend/src/components/Graph.tsx (hunk 10 above)
Comment (IronPan) [reply frontend/src/components/Graph.tsx]: done
